### PR TITLE
[inductor][cpu] Support return_lse and max_scores for compiled FlexAttention on CPU

### PR DESCRIPTION
**Pull request by dumko2001**
## Summary

`torch.compile(flex_attention, ...)` on CPU could not return the two auxiliary
outputs that eager and the CUDA/MPS backends already return:

- `return_lse=True` raised `NotImplementedError("... return_lse is not supported yet")`.
- `max_scores` was rejected by a CPU-only guard in the public API.

So code that asks for the log-sum-exp (needed to stitch attention across chunks,
or for training/inference bookkeeping) or the per-query max score worked in eager
and on GPU but broke the moment it was compiled on CPU. This PR implements both on
the CPU Inductor path so compiled CPU output matches eager, mirroring the existing
MPS support (#187768).

**What was missing.** The CPU flex kernel is emitted through the shared C++ template
autotuner (`CppTemplate` -> `CppTemplateCaller` -> `CppTemplateBuffer`), which only
modeled a single output. LSE/max are written *in place* into extra buffers, i.e. the
kernel mutates them. That "template kernel mutates these buffers" concept existed for
the Triton path but was never surfaced on the CPU path.

**The plumbing fix (shared code).** `mutated_inputs` is now threaded through the CPU
template classes, and `CppTemplateBuffer.get_outputs()` returns the mutation outputs,
exactly like `TritonTemplateBuffer`:

```python
# torch/_inductor/ir.py
class CppTemplateBuffer(TemplateBuffer):
    def get_outputs(self) -> list[Buffer]:
        return [self, *self.mutation_outputs]   # was inherited [self]
```

This matters for correctness, not just tidiness: `_finalize_outputs` reads the aux
buffers to rescale them from log2 to natural log (`lse * ln2`). Without the kernel
being modeled as a *writer*, the scheduler had no ordering edge forcing that rescale
to run after the kernel, so it only worked by incidental program order. For non-flex
C++ templates (GEMM/bmm) `mutation_outputs` is empty, so `get_outputs()` returns
`[self]` and behavior is unchanged.

**The kernel fix.** `flex_cpu.py` allocates float32 `[B, Hq, seq_len_q]` buffers,
appends them to the kernel inputs only when requested, and returns
`(out, logsumexp, max_scores)`. Both CPU templates (the regular one and the
flash-decoding one) now write the stats. The kernel accumulates natural-log softmax
stats but the operator contract is log2, so it emits:

```
lse = (row_max + log(row_sum)) * log2(e)
max = row_max * log2(e)
```

Two masked-row correctness points, both matched against eager:

- **Fully masked query block.** The max/sum accumulators were initialized only inside
  `if (!is_skip_kv)`. A query block with no valid keys skips the kv loop entirely, so
  the aux writes would read stale accumulators. The init is hoisted out of the guard,
  so such rows correctly emit `-inf`.
- **Empty decode partitions.** In flash-decoding, the global max is also the softmax
  normalizer and is clamped to `0` for empty partitions. Reusing it for `max_scores`
  reports `0` even when all real scores are negative. `max_scores` is now reduced from
  the per-partition maxima of partitions that actually saw a token (`exp_sum != 0`),
  giving the true (possibly negative) max, while the softmax normalization is untouched.

## Test Plan

```
python test/inductor/test_flex_attention.py -k test_cpu_aux_outputs_decode_and_masked
python test/inductor/test_flex_attention.py -k test_return_max
```

`test_cpu_aux_outputs_decode_and_masked` compares compiled CPU vs eager for `lse` and
`max_scores` across: the regular and flash-decoding templates, partially and fully
masked query rows/blocks, a decoding case whose true max is negative, `lse`-only, and
legacy `return_lse=True`. `test_return_max` is now enabled on CPU (was skipped) and
passes all score-mod variants. Inference-only compiled CPU flex (no aux) still matches
eager for dense and decoding shapes. `lintrunner` is clean on the changed files.

This change was authored with the assistance of an AI coding assistant; I have reviewed the diff and can explain every line.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by dumko2001** [comment]
@jansel have addressed the claude review

**Comment by dumko2001** [comment]
Hey @jansel , whats the next steps for me here ?

**Comment by jansel** [comment]
@drisspg should review this one

**Comment by jansel** [comment]
ping @drisspg for review

File: torch/_inductor/codegen/cpp_flex_attention_template.py [hunk 1]
```diff
@@ -1086,6 +1140,8 @@ def __init__(
         len_mask_other,
         kernel_input_name_to_buffer,
         block_vars,
+        output_logsumexp=False,
```
Comment (drisspg) [inline torch/_inductor/codegen/cpp_flex_attention_template.py]: do we need default values here?

File: torch/_inductor/codegen/cpp_flex_attention_template.py [hunk 2]
```diff
@@ -1430,6 +1521,10 @@ def render(  # type: ignore[override,return]
             qBlockSize=self.q_block_size,
             template=self,
             output=buf_out,
+            logsumexp=self.logsumexp_buf,
+            max_scores=self.max_scores_buf,
+            output_logsumexp=self.output_logsumexp,
+            output_max=self.output_max,
```
Comment (drisspg) [inline torch/_inductor/codegen/cpp_flex_attention_template.py]: nit if we have max_scores buff we we know we need to output

File: torch/_inductor/codegen/cpp_flex_attention_template.py [hunk 3]
```diff
@@ -1141,17 +1197,41 @@ def get_idx(buf_name):
...
+            self.logsumexp_buf: ir.IRNode | None = self.input_nodes[tail]
+        else:
+            self.logsumexp_buf = None
+        if self.logsumexp_buf is not None:
+            mutated_inputs.append(self.logsumexp_buf)
+        if self.max_scores_buf is not None:
+            mutated_inputs.append(self.max_scores_buf)
+        self.mutated_inputs = mutated_inputs or None
+
     def update_kernel_args(self, kernel_args):
```
Comment (drisspg) [inline torch/_inductor/codegen/cpp_flex_attention_template.py]: I think these buffers should be passed explicitly rather than recovered from the positional tail of `input_nodes`. Please update the constructor and call site accordingly:

```suggestion
        self.logsumexp_buf = logsumexp_buf
        self.max_scores_buf = max_scores_buf
        self.output_logsumexp = logsumexp_buf is not None
        self.output_max = max_scores_buf is not None
        self.mutated_inputs = [
            buf for buf in (logsumexp_buf, max_scores_buf) if buf is not None
        ] or None
```

File: torch/_inductor/codegen/cpp_flex_attention_template.py [hunk 4]
```diff
@@ -1037,6 +1069,28 @@
...
+      for (auto partition_id = 0; partition_id < num_partitions; partition_id++) {
+        if (exp_sum_ptr[i * exp_sum_strideN + j * exp_sum_strideH + partition_id] != 0) {
+          true_max = std::max(true_max, max_logits_ptr[
+              i * max_logits_strideN + j * max_logits_strideH + partition_id]);
+        }
+      }
+      max_scores_data[i * maxStrideB + j * maxStrideH] =
+          global_exp_sum == 0
+              ? -std::numeric_limits<accum_t>::infinity()
+              : true_max * aux_log2e;
```
Comment (drisspg) [inline torch/_inductor/codegen/cpp_flex_attention_template.py]: mCould we make the decode reduction preserve the split-K invariant instead. Empty partitions currently store max=0, so they can make global_max=0; a valid partition with a score such as -200 then gets std::exp(-200) == 0 in float32, producing a zero output and -inf LSE/max.

I think the local stage should store the real partition max and use a separate safe value for the local exponential reduction:

 ```cpp
   max_logits_ptr[...] = partition_max;
   auto softmax_max =
       partition_max == -std::numeric_limits<accum_t>::infinity()
       ? accum_t(0)
       : partition_max;
   exp_reduce_sum_fusion_kernel(..., softmax_max);
   exp_sum_ptr[...] = softmax_max;
 ```

Then the follow  reduction can use exp_sum == 0 ? 0 : std::exp(max_logit- global_max) to avoid exp(-inf- -inf) for fully empty rows. At that point global_max is already the true post-mod max, so this extra true_max loop can go away and max_scores can store global_max * aux_log2e directly.

Comment (drisspg) [inline torch/_inductor/codegen/cpp_flex_attention_template.py]: this is a possible test shat should exercise the differnt semantic for flex decode

> AI-assisted test suggestion:

>
> ```suggestion
>         def neg_score(score, b, h, q_idx, kv_idx):
>             return score * 0.0 - 200.0
> ```
>
> Keep the existing per-head mask:
>
> ```python
> def head_mask(b, h, q_idx, kv_idx):
>     return kv_idx < (512 + h * 256)
> ```
>
> The differing per-head block counts force the CPU decode slow path to create physical KV partitions with no work. Before the fix, those empty partitions store `max=0`, so the valid partitions are rescaled by `exp(-200)` and disappear. The existing eager comparisons for output, LSE, and max should all fail before the fix and pass afterward.
>
> ```bash
> python test/inductor/test_flex_attention.py -k test_cpu_aux_outputs_decode_and_masked
> ```

File: torch/_inductor/codegen/cpp_flex_attention_template.py (hunk 2 above)
Comment (dumko2001) [reply torch/_inductor/codegen/cpp_flex_attention_template.py]: Agreed. The output flag now derives directly from whether `max_scores_buf` is present.

File: torch/_inductor/codegen/cpp_flex_attention_template.py (hunk 4 above)
Comment (dumko2001) [reply torch/_inductor/codegen/cpp_flex_attention_template.py]: Added the `-200` score case; it fails before the fix and passes afterward against eager output, LSE, and max.

Comment (dumko2001) [reply torch/_inductor/codegen/cpp_flex_attention_template.py]: Fixed as suggested: preserve the real partition max, guard empty-partition rescaling, and use the true global max.

File: torch/_inductor/codegen/cpp_flex_attention_template.py (hunk 3 above)
Comment (dumko2001) [reply torch/_inductor/codegen/cpp_flex_attention_template.py]: Done. Both buffers are now passed explicitly, and mutation outputs are built from those arguments.

File: torch/_inductor/codegen/cpp_flex_attention_template.py (hunk 1 above)
Comment (dumko2001) [reply torch/_inductor/codegen/cpp_flex_attention_template.py]: No—the only call site passes the buffer or `None`, so required arguments catch incomplete plumbing.
